runtime/vm/UpcallVMHelpers.cpp

```diff
@@ -114,7 +144,525 @@ native2InterpJavaUpcallD(J9UpcallMetaData *data, void *argsListPointer)
 U_8 * JNICALL
 native2InterpJavaUpcallStruct(J9UpcallMetaData *data, void *argsListPointer)
 {
-	return NULL;
+	UDATA returnValue = native2InterpJavaUpcallImpl(data, argsListPointer);
+	return (U_8 *)returnValue;
+}
+
+/**
+ * @brief Determine the predefined return type against the return signature type
+ * stored in the native signature array of the upcall metadata.
+ *
+ * @param data a pointer to J9UpcallMetaData
+ * @return a U_8 value for the return type
+ */
+static U_8
+getReturnTypeFromMetaData(J9UpcallMetaData *data)
+{
+	J9JavaVM *vm = data->vm;
+	J9VMThread *currentThread = currentVMThread(vm);
+	j9object_t methodType = J9VMOPENJ9INTERNALFOREIGNABIUPCALLMHMETADATA_CALLEETYPE(currentThread,
```

Comment:
is this called while we have VMAccess

Reply:
Moved `getReturnTypeFromMetaData()` to the place after `VM_VMAccess::inlineEnterVMFromJNI(currentThread)` in which case `J9VMOPENJ9INTERNALFOREIGNABIUPCALLMHMETADATA_CALLEETYPE` is called within VMAccess.